File: by-python/py-boj/1629-곱셈.py
def solution(base, e, divisor):
    result = 1

    while e:
        if e & 1:  # e가 홀수면 true, 짝수면 false
            result = result * base % divisor
        e //= 2
        base = base * base % divisor
    return result


def solution_rec(base, e, divisor):
    if e == 0: return 1

    k = solution_rec(base, e // 2, divisor)
    result = k * k % divisor

    if e % 2:
        return base * result % divisor
    else:
        return result


if __name__ == '__main__':
    a, b, c = map(int, input().split())
    print(solution(a, b, c))
    # print(solution_rec(a, b, c))

File: by-kotlin/kt-boj/solutions/1629-곱셈.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.math.BigInteger

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val abc = readLine().split(" ").map { it.toInt() }
    println(solution1629(abc))
}

fun solution1629(abc: List<Int>): BigInteger {
    var result = BigInteger.ONE
    var base = abc[0].toBigInteger()
    var e = abc[1]
    val divisor = abc[2].toBigInteger()

    while (e > 0) {
        if (e % 2 == 1) result = result.multiply(base).mod(divisor)

        e /= 2
        base = base.multiply(base).mod(divisor)
    }

    return result
}
